dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/cloudera/manager/ClouderaAPIHostsTask.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2022-2024 Google LLC
+ * Copyright 2013-2021 CompilerWorks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager;
+
+import com.google.common.io.ByteSink;
+import com.google.edwmigration.dumper.application.dumper.MetadataDumperUsageException;
+import com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager.ClouderaManagerHandle.ClouderaClusterDTO;
+import com.google.edwmigration.dumper.application.dumper.task.TaskRunContext;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import javax.annotation.Nonnull;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.util.EntityUtils;
+
+/**
+ * The tasks dum data from the <a
+ * href="https://archive.cloudera.com/cm7/7.11.3.0/generic/jar/cm_api/apidocs/json_ApiHostList.html">Hosts
+ * API</a> which doesn't contain usage and disk data and collected as a fallback.
+ */
+public class ClouderaAPIHostsTask extends AbstractClouderaManagerTask {
+
+  public ClouderaAPIHostsTask() {
+    super("api-hosts.jsonl");
+  }
+
+  @Override
+  protected Void doRun(
+      TaskRunContext context, @Nonnull ByteSink sink, @Nonnull ClouderaManagerHandle handle)
+      throws Exception {
+    CloseableHttpClient httpClient = handle.getHttpClient();
+    List<ClouderaClusterDTO> clusters = handle.getClusters();
+    if (clusters == null) {
+      throw new MetadataDumperUsageException(
+          "Cloudera clusters must be initialized before hosts dumping.");
+    }
+
+    try (Writer writer = sink.asCharSink(StandardCharsets.UTF_8).openBufferedStream()) {
+      for (ClouderaClusterDTO cluster : clusters) {
+        String hostPerClusterUrl = handle.getApiURI() + "/clusters/" + cluster.getName() + "/hosts";
+
+        try (CloseableHttpResponse hosts = httpClient.execute(new HttpGet(hostPerClusterUrl))) {
+          String json = EntityUtils.toString(hosts.getEntity());
```

Comment:
Have you considered an internal interface / dto to sit between the Cloudera Manager API json response and our dumper output? I guess we should be safe from breaking changes between having a pinned version and the API Client not having updates for 7 years. But it has been common for most dumps to have one: https://github.com/google/dwh-migration-tools/blob/8fc9238b3920ecf8a0d08cfb9e89c43add07f061/dumper/lib-dumper-spi/src/main/java/com/google/edwmigration/dumper/plugin/lib/dumper/spi/BigQueryMetadataDumpFormat.java#L58 

As another example, Hive's output for tables has currently two formats:

- Dumper's internally defined format (https://github.com/google/dwh-migration-tools/blob/8fc9238b3920ecf8a0d08cfb9e89c43add07f061/dumper/lib-dumper-spi/src/main/java/com/google/edwmigration/dumper/plugin/lib/dumper/spi/HiveMetadataDumpFormat.java#L60)
- RAW response from the Thrift call to Hive (https://github.com/google/dwh-migration-tools/blob/8fc9238b3920ecf8a0d08cfb9e89c43add07f061/dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/connector/hive/HiveMetadataConnector.java#L342)

I don't have a particuarly strong opinion on the topic: extra boilerplate vs more control / readability. But I thought it would be good to consider.

Reply:
This comment applies to all Tasks that output a file.

Reply:
I did. In general, the idea of the Dumper is to be a client and to be as simple as possible. All the logic to support different versions of API or fields value should be done on server. It's why I'm trying to avoid strict schemas here and write the responses as is.

There are exceptions:
1. I want to be sure that response is valid json/jsonl format. 
2. I introduces dto `com.google.edwmigration.dumper.application.dumper.connector.cloudera.manager.dto.ApiClusterListDto` for `ClouderaClustersTask`  because this task may do requests to 2 different APIs and  to make sure that success output will be always in single structure `dto` is suitable.